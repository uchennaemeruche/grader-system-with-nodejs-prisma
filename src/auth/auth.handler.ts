import { badImplementation, unauthorized, badRequest } from '@hapi/boom'
import { Request, ResponseToolkit } from '@hapi/hapi'
import { TokenType, UserRole } from '@prisma/client'
import { add } from 'date-fns'
import { TokenService } from './token.service'
import { sign, Algorithm } from 'jsonwebtoken'
import Joi from 'joi'
import { CourseService } from '../course/course.service'

interface LoginInput {
    email: string
}

interface AuthenticateInput {
    email: string
    emailToken: string
}

interface APITokenPayload {
    tokenId: number
}

const apiTokenSchema = Joi.object({
    tokenId: Joi.number().integer().required()
})

export class AuthHandler {
    constructor(
        private tokenService: TokenService,
        private courseService: CourseService,
        private tokenExpiration: number,
        private TOKEN_EXPIRATION_HOURS: number,
        private JWT_ALGORITHM: Algorithm,
        private JWT_SECRET: string
    ) {}
    async login(req: Request, res: ResponseToolkit) {
        const { email } = req.payload as LoginInput
        const emailToken = this.generateToken()

        const tokenExpiresAt = add(new Date(), {
            minutes: this.tokenExpiration
        })

        try {
            const token = await this.tokenService.createToken(
                email,
                tokenExpiresAt,
                TokenType.EMAIL,
                emailToken
            )
            if (token.success) {
                req.server.app.sendEmail(email, emailToken)

                return res
                    .response(
                        'A login token has been sent to your email address'
                    )
                    .code(200)
            }
            return badImplementation(token.error.message)
        } catch (error: any) {
            console.log('ERROR:', error)
            return badImplementation(error.message)
        }
    }

    // authenticate generated email token
    async authenticateToken(req: Request, res: ResponseToolkit) {
        const { email, emailToken } = req.payload as AuthenticateInput
        try {
            const result = await this.tokenService.findToken({ emailToken })
            if (!result.success || !result.emailToken?.valid) {
                return unauthorized()
            }

            // check if token has expired, return 401
            if (result.emailToken.expiration < new Date()) {
                return unauthorized('token expired')
            }

            // check if token matches user email passed in the payload
            if (result.emailToken.user.email !== email) {
                return unauthorized()
            }
            const tokenExpiration = add(new Date(), {
                minutes: this.TOKEN_EXPIRATION_HOURS
            })

            const token = await this.tokenService.createToken(
                email,
                tokenExpiration,
                TokenType.API
            )
            if (!token.success || !token.createdToken) {
                return badRequest()
            }

            await this.tokenService.updateToken(result.emailToken.id, {
                valid: false
            })

            const authToken = this.geneateAuthToken(token.createdToken.id)
            return res.response().code(200).header('Authorization', authToken)
        } catch (error: any) {
            console.log(error)
            return badImplementation(error.message)
        }
    }

    validateApiToken = async (
        decoded: APITokenPayload,
        req: Request,
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        res: ResponseToolkit
    ) => {
        const { tokenId } = decoded
        const { error } = apiTokenSchema.validate(decoded)

        if (error) {
            req.log(['error', 'auth'], `API token error: ${error.message}`)
            return { isValid: false }
        }

        try {
            const token = await this.tokenService.findToken({ id: tokenId })
            if (
                !token.success ||
                !token.emailToken ||
                !token.emailToken.valid
            ) {
                return { isValid: false, error: 'Invalid Token' }
            }

            if (token.emailToken.expiration < new Date()) {
                return { isValid: false, error: 'Token expired' }
            }

            const teacherOf = await this.courseService.findCourseEnrollment({
                userId: token.emailToken.userId,
                role: UserRole.TEACHER
            })
            if (!teacherOf.success || !teacherOf.courseTeacher) {
                return { isValid: false }
            }

            // At this point, the token is valid. Make `userId`, `isAdmin` and `teacherOf` to `credentials` which is available in route handlers via `request.auth.credentials`
            return {
                isValid: true,
                credentials: {
                    tokenId: decoded.tokenId,
                    userId: token.emailToken.userId,
                    isAdmin: token.emailToken.user.isAdmin,
                    // convert teacherOf from an array of objects to an array of numbers
                    teacherOf: teacherOf.courseTeacher.map(
                        ({ courseId }) => courseId
                    )
                }
            }
        } catch (error: any) {
            req.log(['error', 'auth', 'db'], error)
            return { isValid: false }
        }
    }

    private generateToken = () => {
        const token = Math.floor(10000000 + Math.random() * 90000000).toString()
        return token
    }

    private geneateAuthToken = (tokenId: number) => {
        const jwtPayload = { tokenId }

        return sign(jwtPayload, this.JWT_SECRET, {
            noTimestamp: true,
            algorithm: this.JWT_ALGORITHM
        })
    }
}
