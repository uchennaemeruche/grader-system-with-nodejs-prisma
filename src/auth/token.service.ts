import { PrismaClient, TokenType } from '@prisma/client'

export class TokenService {
    constructor(private prisma: PrismaClient) {}

    createToken = async (
        email: string,
        tokenExpiresAt: Date,
        type: TokenType,
        token?: string
    ) => {
        try {
            const createdToken = await this.prisma.token.create({
                data: {
                    type,
                    emailToken: token,
                    expiration: tokenExpiresAt,
                    user: {
                        connectOrCreate: {
                            create: {
                                email
                            },
                            where: {
                                email
                            }
                        }
                    }
                }
            })
            return { success: true, createdToken }
        } catch (error: any) {
            return { success: false, error }
        }
    }
    findToken = async (filter: object) => {
        const emailToken = await this.prisma.token.findUnique({
            where: filter,
            include: {
                user: true
            }
        })
        if (!emailToken) {
            return { success: false, error: new Error('token not found') }
        }
        return { success: true, emailToken }
    }

    updateToken = async (id: number, data = {}) => {
        await this.prisma.token.update({
            where: {
                id
            },
            data
        })
    }
}

// TokenUpdateInput
// TokenWhereUniqueInput
// TokenUpdateArgs
