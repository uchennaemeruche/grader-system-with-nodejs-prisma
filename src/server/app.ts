import Hapi, { Server } from '@hapi/hapi'
import { emailPlugin } from '../plugins/email'
import { prismaPlugin } from '../plugins/prisma'
import usersPlugin from '../users/users.route'

export class AppServer {
    private server: Server
    constructor(port: number, host: string) {
        this.server = Hapi.server({ port, host })
        this.initRoutes()
    }

    private async initRoutes() {
        this.server.route({
            method: 'GET',
            path: '/',
            handler: (_, h) => {
                return h.response({ up: true }).code(200)
            }
        })
    }

    public async createServer() {
        await this.server.register([prismaPlugin, usersPlugin, emailPlugin])
        await this.server.initialize()

        return this.server
    }

    public async start(server: Server) {
        server.start()
        console.log(
            `App Listening on ${server.settings.host}:${server.settings.port}`
        )

        process.on('unhandledRejection', (err) => {
            console.error('unhandledRejection')
            console.error(err)
            process.exit(1)
        })
    }
}
