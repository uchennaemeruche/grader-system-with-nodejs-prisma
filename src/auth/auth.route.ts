import { Plugin, Server } from '@hapi/hapi'
import Joi from 'joi'
import { AuthHandler } from './auth.handler'

const secret = process.env.JWT_SECRET || 'SUPER_SECRET_JWT_SECRET'
const handler = new AuthHandler(10, 12, 'HS256', secret)

export const authPlugin: Plugin<null> = {
    name: 'app/auth',
    dependencies: ['prisma', 'hapi-auth-jwt2', 'app/email'],
    register: async function (server: Server) {
        server.route({
            method: 'POST',
            path: '/auth/login',
            handler: handler.login,
            options: {
                auth: false,
                validate: {
                    payload: Joi.object({
                        email: Joi.string().email().required()
                    })
                }
            }
        })
        server.route({
            method: 'POST',
            path: '/auth/authenticate',
            handler: handler.authenticate,
            options: {
                auth: false,
                validate: {
                    payload: Joi.object({
                        email: Joi.string().email().required(),
                        emailToken: Joi.string().required()
                    })
                }
            }
        })
    }
}
