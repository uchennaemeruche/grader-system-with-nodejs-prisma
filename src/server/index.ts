import { AppServer } from './app'
import dotenv from 'dotenv'

dotenv.config()

const PORT = Number(process.env.PORT) | 4000
const app = new AppServer(PORT, 'localhost')

app.createServer().then(app.start)
