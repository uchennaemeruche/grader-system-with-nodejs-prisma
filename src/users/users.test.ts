import { Server } from '@hapi/hapi'
import { API_AUTH_STATEGY } from '../auth/auth.route'
import { AppServer } from '../server/app'
import { createUserCredentials } from '../test/test.helper'

describe('Users Test', () => {
    let server: Server
    beforeAll(async () => {
        server = await new AppServer(3000, 'localhost').createServer()
    })

    afterAll(async () => {
        await server.stop()
    })
    let userId: number
    describe('Create User', () => {
        it('creates a user and returns 201', async () => {
            const response = await server.inject({
                method: 'POST',
                url: '/users',
                payload: {
                    name: 'test--name',
                    email: `test-${Date.now()}@example.io`,
                    social: {
                        twitter: 'giveittojest'
                    }
                },
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        false
                    )
                }
            })
            userId = JSON.parse(response.payload)?.id
            expect(response.statusCode).toEqual(201)
            expect(response.result).toHaveProperty('id')
            expect(
                typeof JSON.parse(response.payload)?.id === 'number'
            ).toBeTruthy()
        })
        it('returns an error for an incomplete input payload', async () => {
            const response = await server.inject({
                method: 'POST',
                url: '/users',
                payload: {
                    email: `test-${Date.now()}@example.io`,
                    social: {
                        twitter: 'giveittojest'
                    }
                },
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        false
                    )
                }
            })
            expect(response.statusCode).toEqual(400)
            expect(response.result).toHaveProperty('error')
            expect(response.result).toHaveProperty('message')
        })
    })

    describe('Get Users', () => {
        it('Returns a list of created users', async () => {
            const response = await server.inject({
                method: 'GET',
                url: '/users',
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        false
                    )
                }
            })

            const result = response.result as Array<object>

            expect(response.statusCode).toEqual(200)
            expect(Array.isArray(['response.result'])).toBe(true)
            expect(result.length).toBeGreaterThanOrEqual(1)
            expect(result[0]).toMatchSnapshot({
                id: expect.any(Number),
                email: expect.any(String),
                name: expect.any(String),
                social: expect.any(Object),
                isAdmin: expect.any(Boolean)
            })
        })
        it('Returns a single user when the requesting Non-Admin User ID equals the requested user ID', async () => {
            const credentials = await createUserCredentials(
                server.app.prisma,
                false
            )
            const response = await server.inject({
                method: 'GET',
                url: `/users/${credentials.userId}`,
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials
                }
            })
            expect(response.statusCode).toEqual(200)
            expect(response.result).toMatchSnapshot({
                id: expect.any(Number),
                email: expect.any(String),
                name: expect.any(String),
                social: expect.any(Object),
                isAdmin: expect.any(Boolean)
            })
        })
        it('Returns a single user when the requesting User isAdmin', async () => {
            const credentials = await createUserCredentials(
                server.app.prisma,
                true
            )
            const response = await server.inject({
                method: 'GET',
                url: `/users/${userId}`,
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials
                }
            })
            expect(response.statusCode).toEqual(200)
            expect(response.result).toMatchSnapshot({
                id: expect.any(Number),
                email: expect.any(String),
                name: expect.any(String),
                social: expect.any(Object),
                isAdmin: expect.any(Boolean)
            })
        })
        it('Returns a 403 Error when the requesting user is not equals to user ID', async () => {
            const credentials = await createUserCredentials(
                server.app.prisma,
                false
            )
            const response = await server.inject({
                method: 'GET',
                url: `/users/${userId}`,
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials
                }
            })

            expect(response.statusCode).toEqual(403)
            expect(response.result).toHaveProperty('error')
        })
        it('Returns 404 for a wrong or non-existing user', async () => {
            const response = await server.inject({
                method: 'GET',
                url: '/users/1000',
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        true
                    )
                }
            })

            expect(response.statusCode).toEqual(404)
        })
        it('Returns a 400 validation error for an invalid param', async () => {
            const response = await server.inject({
                method: 'GET',
                url: '/users/aaksllss',
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        true
                    )
                }
            })
            expect(response.statusCode).toEqual(400)
            expect(response.result).toHaveProperty('error')
        })
    })
    describe('Update User', () => {
        it('fails and returns 400 error given an invalid userId parameter', async () => {
            const response = await server.inject({
                method: 'PUT',
                url: `/users/jfjjkdkd`,
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        true
                    )
                }
            })
            expect(response.statusCode).toEqual(400)
        })
        it('returns 404 for non-existing userId', async () => {
            const response = await server.inject({
                method: 'PUT',
                url: `/users/1`,
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        true
                    )
                }
            })
            expect(response.statusCode).toEqual(400)
        })
        it('updates a user when requesting User Id equals ID of user to be updated', async () => {
            const updatedName = 'Uchechukwu Emeruche'
            const credentials = await createUserCredentials(
                server.app.prisma,
                false
            )
            const response = await server.inject({
                method: 'PUT',
                url: `/users/${credentials.userId}`,
                payload: {
                    name: updatedName
                },
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials
                }
            })
            expect(response.statusCode).toEqual(200)
            const user = JSON.parse(response.payload)
            expect(user.name).toEqual(updatedName)
        })
        it('updates a user when requesting User equals admin', async () => {
            const updatedName = 'Uchechukwu Emeruche'
            const response = await server.inject({
                method: 'PUT',
                url: `/users/${userId}`,
                payload: {
                    name: updatedName
                },
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        true
                    )
                }
            })
            expect(response.statusCode).toEqual(200)
            const user = JSON.parse(response.payload)
            expect(user.name).toEqual(updatedName)
        })
    })

    describe('Delete User', () => {
        it('fails and returns 400 error with invalid userId parameter ', async () => {
            const response = await server.inject({
                method: 'DELETE',
                url: '/users/jsj002',
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        true
                    )
                }
            })
            expect(response.statusCode).toEqual(400)
        })
        it('Deletes a user when requesting User Id equals ID of user to be deleted', async () => {
            const credentials = await createUserCredentials(
                server.app.prisma,
                false
            )
            const response = await server.inject({
                method: 'DELETE',
                url: `/users/${credentials.userId}`,
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: credentials
                }
            })
            expect(response.statusCode).toEqual(204)
        })
        it('Deletes a user when requesting User is admin', async () => {
            const response = await server.inject({
                method: 'DELETE',
                url: `/users/${userId}`,
                auth: {
                    strategy: API_AUTH_STATEGY,
                    credentials: await createUserCredentials(
                        server.app.prisma,
                        true
                    )
                }
            })
            expect(response.statusCode).toEqual(204)
        })
    })
})
