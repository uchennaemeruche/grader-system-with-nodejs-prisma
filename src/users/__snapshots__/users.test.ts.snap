// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Test Get Users Returns a list of created users 1`] = `
{
  "email": Any<String>,
  "id": Any<Number>,
  "isAdmin": Any<Boolean>,
  "name": Any<String>,
  "social": Any<Object>,
}
`;

exports[`Users Test Get Users Returns a single user when the requesting Non-Admin User ID equals the requested user ID 1`] = `
{
  "email": Any<String>,
  "id": Any<Number>,
  "isAdmin": Any<Boolean>,
  "name": Any<String>,
  "social": Any<Object>,
}
`;

exports[`Users Test Get Users Returns any single user when the requesting User isAdmin 1`] = `
{
  "email": Any<String>,
  "id": Any<Number>,
  "isAdmin": Any<Boolean>,
  "name": Any<String>,
  "social": Any<Object>,
}
`;
